userscript-preferences.js

```diff
@@ -0,0 +1,230 @@
+The key 'localStorage' is optional. This will be used as a key in the
+browser's localStorage. If provided, any 'value' argument will be ignored
+
+Example user-code:
+
+if ( window.location.pathname === "/preferences" ) {
+    add_userscript_settings(
+        "My Heading",
+        [
+            {
+                label: "Text area",
+                type: "text",
+                value: "value 1",
+                callback: function(value) {
+                    alert(value);
+                }
+            },
+            {
+                label: "Checkbox",
+                type: "checkbox",
+                value: false,
+                callback: function(value) {
+                    alert(value);
+                }
+            },
+            {
+                label: "Radio buttons",
+                type: "radio",
+                value: "one",
+                options: ["one", "two"],
+                callback: function(value) {
+                    alert(value);
+                }
+            }
+        ]
+    );
+}
+
+*/
+
+var core_prefs;
+
+function add_userscript_settings(def) {
+    "use strict";
+
+    if ( "/preferences" !== window.location.pathname ) {
+        return;
+    }
+
+    let stored_defaults = {};
+    let stored_defaults_str = localStorage.getItem( def.key );
+    if ( stored_defaults_str ) {
+        try {
+            stored_defaults = JSON.parse( stored_defaults_str );
+        }
+        catch (error) {
+            // console.log("caught error");
+            // console.log(error);
+        }
+    }
+    let def_defaults = {};
+    if ( def.hasOwnProperty("defaults") ) {
+        def_defaults = def.defaults;
+    }
+
+
+    if ( !core_prefs ) {
+        core_prefs = $("#preferences");
+    }
+
+    let label = def.label || def.key;
+    core_prefs.append( `<h2>UserScript: ${label}</h2>` );
+
+    for ( let i=0; i<def.options.length; i++ ) {
+        let pref = def.options[i];
+        let dl = $("<dl></dl>");
+        let dd = $("<dd></dd>");
+
+        core_prefs.append(dl);
+        let this_label = pref.label || pref.key;
+        dl.append(
+            `<dt>${this_label}</dt>`,
+            dd
+        );
+        let this_value;
+        if ( stored_defaults.hasOwnProperty( pref.key ) ) {
+            this_value = stored_defaults[ pref.key ];
+        }
+        else if ( def_defaults.hasOwnProperty( pref.key ) ) {
+            this_value = def_defaults[ pref.key ];
+        }
+        let this_id = def.key + "_" + pref.key;
+
+        switch ( pref.type ) {
+            case "text":
+                dd.append(
+                    `<input data-userscript-pref='${this_id}' type='text' value='${this_value}' />`
+                );
+                break;
+            case "checkbox":
+                let checked = this_value ? "checked=checked" : "";
+                dd.append(
+                    `<input data-userscript-pref='${this_id}' type='checkbox' value=1 ${checked}/>`
```

Comment:
For simple boolean checkboxes: If you let the user click/tap on the item's `this_label` to change the setting, the larger tap target would make it easier for mobile users to use.  Attaching a click() handler would be one approach; alternately, could you get the same effect by adding a `<label>...</label>` tag surrounding both `this_label` & the `<input ... />` tag, like you do for the array checkboxes?)

Reply:
I'm certainly heading towards your suggestion of using the same On/Off link UI that the game uses - and this would solve the target-size issue.
I wouldn't be keen on saving the result of `onchange` for text boxes (saving after each character change, and potentially calling a callback too), so maybe either automatically save after losing focus or adding a `save` button after each text box if the value changes?